xline-client/src/clients/kv.rs

```diff
@@ -27,4 +31,762 @@ impl KvClient {
+    pub fn with_range(mut self, end: impl Into<Vec<u8>>) -> Self {
+        self.0.range_end = end.into();
+        self
+    }
+
+    /// Sets the comparison to scan all keys prefixed by the key.
+    #[inline]
+    #[must_use]
+    pub fn with_prefix(mut self) -> Self {
+        self.0.range_end = KeyRange::get_prefix(&self.0.key);
+        self
+    }
+}
+
+/// Transaction operation.
+#[derive(Debug, Clone)]
+pub struct TxnOp {
+    /// The inner txn op request
+    inner: xlineapi::Request,
+}
+
+impl TxnOp {
+    /// `Put` operation.
+    #[inline]
+    #[must_use]
+    pub fn put(reqeust: PutRequest) -> Self {
+        TxnOp {
+            inner: xlineapi::Request::RequestPut(reqeust.into()),
+        }
+    }
+
+    /// `Get` operation.
+    #[inline]
+    #[must_use]
+    pub fn range(request: RangeRequest) -> Self {
+        TxnOp {
+            inner: xlineapi::Request::RequestRange(request.into()),
+        }
+    }
+
+    /// `Delete` operation.
+    #[inline]
+    #[must_use]
+    pub fn delete(request: DeleteRangeRequest) -> Self {
+        TxnOp {
+            inner: xlineapi::Request::RequestDeleteRange(request.into()),
+        }
+    }
+
+    /// `Txn` operation.
+    #[inline]
+    #[must_use]
+    pub fn txn(txn: Txn) -> Self {
+        TxnOp {
+            inner: xlineapi::Request::RequestTxn(txn.into()),
+        }
+    }
+}
+
+impl From<TxnOp> for xlineapi::Request {
+    #[inline]
+    fn from(op: TxnOp) -> Self {
+        op.inner
+    }
+}
+
+/// Transaction of multiple operations.
+#[derive(Debug)]
+pub struct Txn {
+    /// the inner txn request
+    req: xlineapi::TxnRequest,
+    /// if `when` have be set
+    c_when: bool,
+    /// if `then` have be set
+    c_then: bool,
+    /// if `else` have be set
+    c_else: bool,
+}
+
+impl Txn {
+    /// Creates a new transaction.
+    #[inline]
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            req: xlineapi::TxnRequest {
+                compare: Vec::new(),
+                success: Vec::new(),
+                failure: Vec::new(),
+            },
+            c_when: false,
+            c_then: false,
+            c_else: false,
+        }
+    }
+
+    /// Takes a list of comparison. If all comparisons passed in succeed,
+    /// the operations passed into `and_then()` will be executed. Or the operations
+    /// passed into `or_else()` will be executed.
+    ///
+    /// # Panics
+    ///
+    /// panics if `when` is called twice or called after `when` or called after `or_else`
+    #[inline]
+    #[must_use]
+    pub fn when(mut self, compares: impl Into<Vec<Compare>>) -> Self {
+        assert!(!self.c_when, "cannot call when twice");
+        assert!(!self.c_then, "cannot call when after and_then");
+        assert!(!self.c_else, "cannot call when after or_else");
+
+        let compares_vec: Vec<Compare> = compares.into();
+        self.c_when = true;
+        self.req.compare = compares_vec.iter().map(|c| c.clone().0).collect();
```

Comment:
Please use `into_iter` then we don't need to clone it.

Reply:
done